File: csharp/src/Vanni.Idmix/Codec.cs
using System.Text;

namespace Vanni.Idmix;

/// <summary>包级文本编解码辅助与内置 Codec 实现。</summary>
public static class Codec
{
    private static readonly Lazy<ICodec> DefaultCodec = new(() => new RadixCodec(IdMix.DefaultAlphabet));

    /// <summary>将任意二进制编码为文本；codec 不传时使用默认 RadixCodec。</summary>
    public static string EncodeBytes(byte[] data, ICodec? codec = null) =>
        (codec ?? DefaultCodec.Value).Encode(data);

    /// <summary>将文本还原为二进制；codec 不传时使用默认 RadixCodec。</summary>
    public static byte[] DecodeString(string s, ICodec? codec = null) =>
        (codec ?? DefaultCodec.Value).Decode(s);
}

/// <summary>标准 Base64 编解码器。</summary>
public sealed class Base64Codec : ICodec
{
    public static Base64Codec Instance { get; } = new();

    public string Encode(byte[] data) => Convert.ToBase64String(data);

    public byte[] Decode(string s) => Convert.FromBase64String(s);
}

/// <summary>由函数实现的 Codec，便于包装 AES/XOR 等自定义逻辑。</summary>
public sealed class FuncCodec : ICodec
{
    public Func<byte[], string>? EncodeFn { get; init; }
    public Func<string, byte[]>? DecodeFn { get; init; }

    public string Encode(byte[] data)
    {
        if (EncodeFn == null) throw new InvalidOperationException("codec function is nil");
        return EncodeFn(data);
    }

    public byte[] Decode(string s)
    {
        if (DecodeFn == null) throw new InvalidOperationException("codec function is nil");
        return DecodeFn(s);
    }
}

File: java/src/main/java/io/github/vannifan/idmix/Codec.java
package io.github.vannifan.idmix;

import java.util.Base64;
import java.util.function.Function;

/** Package-level text codec helpers and built-in implementations. */
public final class Codec {
    private static final ICodec DEFAULT_CODEC = new RadixCodec(IdMix.DEFAULT_ALPHABET);

    private Codec() {}

    public static String encodeBytes(byte[] data) {
        return encodeBytes(data, null);
    }

    public static String encodeBytes(byte[] data, ICodec codec) {
        return (codec != null ? codec : DEFAULT_CODEC).encode(data);
    }

    public static byte[] decodeString(String s) {
        return decodeString(s, null);
    }

    public static byte[] decodeString(String s, ICodec codec) {
        return (codec != null ? codec : DEFAULT_CODEC).decode(s);
    }

    /** Standard Base64 codec. */
    public static final class Base64Codec implements ICodec {
        public static final Base64Codec INSTANCE = new Base64Codec();

        private Base64Codec() {}

        @Override
        public String encode(byte[] data) {
            return Base64.getEncoder().encodeToString(data);
        }

        @Override
        public byte[] decode(String s) {
            return Base64.getDecoder().decode(s);
        }
    }

    /** Function-based codec for custom encryption/XOR logic. */
    public static final class FuncCodec implements ICodec {
        private final Function<byte[], String> encodeFn;
        private final Function<String, byte[]> decodeFn;

        public FuncCodec(Function<byte[], String> encodeFn, Function<String, byte[]> decodeFn) {
            this.encodeFn = encodeFn;
            this.decodeFn = decodeFn;
        }

        @Override
        public String encode(byte[] data) {
            if (encodeFn == null) throw new IllegalStateException("codec function is nil");
            return encodeFn.apply(data);
        }

        @Override
        public byte[] decode(String s) {
            if (decodeFn == null) throw new IllegalStateException("codec function is nil");
            return decodeFn.apply(s);
        }
    }
}
